Fix Several Components Not Disabled During Execution

Xiao-zhen-Liu opened the pull request:
This PR fixes a bug where the following functionalities are not disabled when `workflowModificatonEnabled` is `true`:

- Disable operator
- Cache operator
- Auto Layout
- Delete All
- Import Workflow
- Add Comment
- Restore to a Version
- Python UDF's Edit button
- Python UDF code editor
- Context Menu items

These bugs are introduced by #1674's deletion of "lock" in previous shared editing implementation.

For other components, I use `WorkflowActionService`'s flag. For Python UDF's readonly status I use formly's status, since we allow temporarily enabling modification of properties during execution.

Added back the "readonly" button for UDF, and used a workaround similar to #1688 to fix the unit test. It would look like below:
![Screen Recording 2022-12-08 at 11 30 20](https://user-images.githubusercontent.com/36582710/206551162-d2727068-2b65-494f-a8f6-a494eea3adc2.gif)

There is one more complexity for version display: since when displaying a version, workflow modification is always disabled, we cannot use this flag to determine whether restore version should be disabled. This has never been handled properly. For now I added a temporary flag to remember the workflow action disabled status before displaying a particular version.

Yicong-Huang commented:
I remember @yunyad has occurred some issue with FormlyControl and has a workaround. Maybe it can apply here so we can update the tests instead of turning it off. Even though the test case may not doing much, it at least can testify if the component can be initiated etc, so better to keep the tests if possible.

Xiao-zhen-Liu commented:
> I remember @yunyad has occurred some issue with FormlyControl and has a workaround. Maybe it can apply here so we can update the tests instead of turning it off. Even though the test case may not doing much, it at least can testify if the component can be initiated etc, so better to keep the tests if possible.

Yes I fixed the issue by not using `this.formControl` directly and enforcing `undefined` check for `this.field` before subscribing.

I also added back the "View Code" button when editing is disabled.

Yicong-Huang commented on core/new-gui/src/app/workspace/component/code-editor-dialog/code-editor-dialog.component.ts:
```diff
@@ -36,20 +37,21 @@ export class CodeEditorDialogComponent implements AfterViewInit, SafeStyle, OnDe
     private dialogRef: MatDialogRef<CodeEditorDialogComponent>,
-    @Inject(MAT_DIALOG_DATA) code: any,
+    @Inject(MAT_DIALOG_DATA) formControl: FormControl,
     private workflowActionService: WorkflowActionService,
     public coeditorPresenceService: CoeditorPresenceService
   ) {
-    this.code = code;
+    this.formControl = formControl;
```
ok.. I see you intention is to use formControl's status to change editor to be editable or not... but I think formControl should not be the source of truth for this behavior. We should have a workflow editable or not boolean somewhere, maybe in WorkflowActionService, here the editor should subscribe to that boolean's state, which is the single source of truth. What do you think?

Xiao-zhen-Liu replied:
No we cannot because we allow temporarily enabling of only properties of an operator to debug the operator which is separate from the `checkWorkflowModificationEnabled()` in workflowActionService. The temporarily enabled status is only in the form component. I explained this in the description :)

Yicong-Huang replied:
I did not fully understand that part from your description. No I have better understanding with https://github.com/Texera/texera/pull/1763#discussion_r1043768712. We can continue discussion there.

Yicong-Huang commented on core/new-gui/src/app/workspace/component/codearea-custom-template/codearea-custom-template.component.ts:
```diff
@@ -20,20 +20,36 @@ import { CoeditorPresenceService } from "../../service/workflow-graph/model/coed
-export class CodeareaCustomTemplateComponent extends FieldType {
+export class CodeareaCustomTemplateComponent extends FieldType<any> implements AfterViewInit {
   dialogRef: MatDialogRef<CodeEditorDialogComponent> | undefined;
+  disabled: boolean = false;
```
I understand we want to disable editing during a workflow execution. However, this particular boolean is describing the CodeareaCustomTemplateComponent, which is never fully "disabled" (i.e., greyed out, not able to click). It shows "View Code" in readonly mode and "Edit Code" in normal mode. The boolean name disabled is confusing here and does not reflect the component's behavior, better change it to be more meaningful.

Xiao-zhen-Liu replied:
Done.

Yicong-Huang commented on core/new-gui/src/app/workspace/component/codearea-custom-template/codearea-custom-template.component.ts:
```diff
@@ -20,20 +20,36 @@ import { CoeditorPresenceService } from "../../service/workflow-graph/model/coed
+   */
+  handleDisabled(): void {
+    if (this.field !== undefined)
+      this.field.formControl.statusChanges.pipe(untilDestroyed(this)).subscribe(() => {
```
I again think this should listen to WorkflowActionService, rather than formControl's status change.

Yicong-Huang replied:
ok, understood. so unlock for logic change is done locally, without going through `WorkflowActionService`. I have a feeling that it's better to manage those disabled states within `WorkflowActionService`, but I do not have a better suggestion for now. Can you leave comments/todo in code so we aware of this behavior, and to revisit in the future?

Xiao-zhen-Liu replied:
Added.

Yicong-Huang reviewed (approved):
Looks good to me. 

One general question, are all the components that ought to be disabled during execution handled properly now? For example, how about buttons on navigation bar that could trigger writing to canvas, such as,`delete all`, `import workflow`, `add comment`, `restore to a version`, etc, as well as the buttons from the right click menus, are they disabled as well? If not, what is our plan for them?

Xiao-zhen-Liu commented:
> Looks good to me.
> 
> One general question, are all the components that ought to be disabled during execution handled properly now? For example, how about buttons on navigation bar that could trigger writing to canvas, such as,`delete all`, `import workflow`, `add comment`, `restore to a version`, etc, as well as the buttons from the right click menus, are they disabled as well? If not, what is our plan for them?

I added disabling of these during execution too. These are all that was impacted by #1674.

I also realized one more complexity for version display: since when displaying a version, workflow modification is always disabled, we cannot use this flag to determine whether restore version should be disabled. This has never been handled properly. For now I added a temporary flag to remember the workflow action disabled status before displaying a particular version.

I also added such support in other items in the right click menus (delete, cut, and paste), which were not handled properly in #1647.

These should be all of them.

Yicong-Huang commented:
> > Looks good to me.
> > One general question, are all the components that ought to be disabled during execution handled properly now? For example, how about buttons on navigation bar that could trigger writing to canvas, such as,`delete all`, `import workflow`, `add comment`, `restore to a version`, etc, as well as the buttons from the right click menus, are they disabled as well? If not, what is our plan for them?
> 
> I added disabling of these during execution too. These are all that was impacted by #1674.
> 
> I also realized one more complexity for version display: since when displaying a version, workflow modification is always disabled, we cannot use this flag to determine whether restore version should be disabled. This has never been handled properly. For now I added a temporary flag to remember the workflow action disabled status before displaying a particular version.
> 
> I also added such support in other items in the right click menus (delete, cut, and paste), which were not handled properly in #1647.
> 
> These should be all of them.

Thanks! should I take another look of the new changes?

Yicong-Huang commented:
> > I added disabling of these during execution too. These are all that was impacted by #1674.
> > I also realized one more complexity for version display: since when displaying a version, workflow modification is always disabled, we cannot use this flag to determine whether restore version should be disabled. This has never been handled properly. For now I added a temporary flag to remember the workflow action disabled status before displaying a particular version.
> > I also added such support in other items in the right click menus (delete, cut, and paste), which were not handled properly in #1647.
> > These should be all of them.
> 
> Thanks! should I take another look of the new changes?

Also, can you please add those new changes to the PR description?

Xiao-zhen-Liu commented:
> > > Looks good to me.
> > > One general question, are all the components that ought to be disabled during execution handled properly now? For example, how about buttons on navigation bar that could trigger writing to canvas, such as,`delete all`, `import workflow`, `add comment`, `restore to a version`, etc, as well as the buttons from the right click menus, are they disabled as well? If not, what is our plan for them?
> > 
> > 
> > I added disabling of these during execution too. These are all that was impacted by #1674.
> > I also realized one more complexity for version display: since when displaying a version, workflow modification is always disabled, we cannot use this flag to determine whether restore version should be disabled. This has never been handled properly. For now I added a temporary flag to remember the workflow action disabled status before displaying a particular version.
> > I also added such support in other items in the right click menus (delete, cut, and paste), which were not handled properly in #1647.
> > These should be all of them.
> 
> Thanks! should I take another look of the new changes?

@Yicong-Huang please do, thanks.

Yicong-Huang commented on core/new-gui/src/app/workspace/component/code-editor-dialog/code-editor-dialog.component.ts:
```diff
@@ -36,20 +37,21 @@ export class CodeEditorDialogComponent implements AfterViewInit, SafeStyle, OnDe
   loaded: boolean = false;
 
+  private formControl: FormControl;
   private ytext?: YText;
```
I know this is unrelated for this PR, but since you removed the code variable now, I believe this variable is the actual code for the editor. can we name it as "code" with type "YText"?

Xiao-zhen-Liu replied:
Done.

Yicong-Huang commented on core/new-gui/src/app/workspace/component/codearea-custom-template/codearea-custom-template.component.ts:
```diff
@@ -20,20 +20,40 @@ import { CoeditorPresenceService } from "../../service/workflow-graph/model/coed
+   *  is enabled, which can only be accessed via <code>formControl</code>. Might need a more unified solution.
+   */
+  handleReadonlyStatusChange(): void {
+    if (this.field !== undefined)
```
It's Better to check for undefined and return early explicitly. Please swap the of condition.

Xiao-zhen-Liu replied:
Done.

Yicong-Huang reviewed (approved):
LGTM. Left some minor comments.
